File: Tries/211. Design Add and Search Words Data Structure.py
class WordDictionary:

    def __init__(self):
        self.endSymbol = "*"
        self.root = {}
        
        
    def addWord(self, word: str) -> None:
        currentNode = self.root
        for index in range(len(word)):
            currentCharacter = word[index]
            if currentCharacter not in currentNode:
                currentNode[currentCharacter] = {}
            currentNode = currentNode[currentCharacter]
            
        currentNode[self.endSymbol] = True
        
    def search(self, word: str) -> bool:
        
        def dfs(root, currentWord):
            if root == True:
                return False 
            currentNode = root
            for index in range(len(currentWord)):
                currentCharacter = currentWord[index]
                if currentCharacter == ".":
                    newWord = currentWord[index+1:]
                    for char in currentNode:
                        newRoot = currentNode[char]
                        if dfs(newRoot, newWord):
                            return True 
                    return False 
                elif currentCharacter not in currentNode:
                    return False 
                currentNode = currentNode[currentCharacter]
            if type(currentNode) == bool :
                return False 
            elif type(currentNode) == dict :
                return "*" in currentNode
                
        
        return dfs(self.root, word)

File: Tries/211. Design Add and Search Words Data Structure.cpp
class WordDictionary {
    struct TrieNode {
        unordered_map<char, TrieNode*> leaves;
        bool is_string = false;
    };
    TrieNode* root = nullptr;
public:
    WordDictionary() {
        root = new TrieNode();

    }

    void addWord(string word) {
        TrieNode* ptr = root;
        for (char& c: word) {
            if (ptr->leaves.count(c) == 0) {
                ptr->leaves[c] = new TrieNode();
            }
            ptr = ptr->leaves[c];
        }
        ptr->is_string = true;

    }
    bool doesExist(string& word, int index, TrieNode* ptr) {
        if (index == word.length()) {
            return ptr->is_string;
        }

        char cur = word[index];
        if (cur == '.') {
            // wild card
            for (auto it = ptr->leaves.begin(); it != ptr->leaves.end(); it++) {
                if (doesExist(word, index + 1, it->second)) return true;
            }

        } else {
            if (ptr->leaves.find(cur) != ptr->leaves.end()) {
                return doesExist(word, index + 1, ptr->leaves[cur]);

            }
        }
        return false;
    }

    bool search(string word) {
        return doesExist(word, 0, root);
    }
};